Remove caller_id requirement from schedule(interval_secs=#)

com4 opened the pull request:
- Serializes full argument list for messages and hashes that to uniquely
  identify jobs (fixes #73)
- Removes unused `raw_jobs` part of code in `load_jobs`
- Removes unused `cron_hash` method
- Adds `schedule`/`unschedule` helper functions (creates the socket for the user)
- Add coveralls support for a fancy test coverage percent badge. https://coveralls.io/github/enderlabs/eventmq

com4 commented:
For now, just review this. I will need to write a utility to convert all the job hashes from the old style to the new style before we merge and deploy.

synic commented on eventmq/scheduler.py:
```diff
@@ -417,32 +387,36 @@ def on_heartbeat(self, msgid, message):
         # Hash the sorted, immutable set of items in our identifying dict
-        schedule_hash = str(hash(tuple(frozenset(sorted(
-            schedule_hash_items.items())))))
+        schedule_hash = str(hash(schedule_hash_items))
```
Probably needs a way to ignore certain arguments, because passing things like `track_uuid` that change every time will screw it up, no?

com4 replied:
No, I think passing some argument that is designed to change into a scheduled task is a bad idea. 

`track_uuid` will be stored as an argument and be passed to all subsequent calls of the task. This defeats the purpose of a tracking id in this case. Your search query will be "show me all the things that happened for this request" and you will get a bunch of stuff that has nothing to do with the request. What you probably want instead is to see that something was scheduled and not every single call that happened because of the call to `schedule`.

I considered allowing the caller to pass an id for the task in the schedule function.  I will add in that in if you really want to be passing that id in.

synic replied:
Fair enough, just something to think about.

sideshowdave7 replied:
I agree, tracking should be the job of the framework and not end developer in this case

sideshowdave7 replied:
It shouldn't be something you consciously put in there, it should be available automatically (which it is with eventmq)

sideshowdave7 replied:
are we certain without sorting we are gaurenteed the same hash everytime?

com4 replied:
When I was writing the code to track eventmq jobs, I was going to repurpose the id of the message as the tracking id. Sadly I was pulled off before it was complete.

I still think that using the message id would be the best way to track jobs. We can expose a way to provide the message id to the `defer_job` and `schedule` helpers if that seems like a good idea.

sideshowdave7 replied:
agreed

com4 replied:
`sorted` doesn't sort dictionaries so `sort_keys` is used on the json dumps call

sideshowdave7 replied:
Ah i missed that instead you are serializing the dict instead of using the items list -> serializing will be much better for sure.

com4 replied:
I added a function (test_scheduler.py:32) to test that reordering the items in a dictionary doesn't change the value of the hash.

sideshowdave7 replied:
yea it looks solid

com4 commented:
Here is the script to convert the hashes for scheduled jobs. It loops through everything in redis, unschedules the jobs (and for good measure deletes it from redis), then loops through all the jobs again and re-calls the schedule function.  Once this PR is merged I will test it out on ninja a couple of times before we try it on production.

The output it produces should be enough to fix any problems that we run into if that happens.
```python
"""
This script is used to update the scheduler hashes from the old style
(caller_id) to the new style (func signature).
"""
import json

from eventmq.scheduler import Scheduler
from eventmq import schedule, unschedule


def old_schedule_hash(message):
    """
    Create a unique identifier for this message for storing
    and referencing later

    Args:
        message (str): The serialized message passed to the scheduler

    Returns:
        int: unique hash for the job
    """
    # Get the job portion of the message
    msg = json.loads(message[3])[1]

    schedule_hash_items = {'caller_id': msg['class_args'][0],
                           'path': msg['path'],
                           'callable': msg['callable']}

    # Hash the sorted, immutable set of items in our identifying dict
    schedule_hash = str(hash(tuple(frozenset(sorted(
        schedule_hash_items.items())))))

    return schedule_hash


if __name__ == "__main__":
    s = Scheduler()
    r = s.redis_server

    interval_job_list = r.lrange('interval_jobs', 0, -1)

    jobs = []
    for old_hash in interval_job_list:
        job_msg = json.loads(r.get(old_hash))
        job = json.loads(job_msg[3])[1]

        func = '.'.join(job['path'], job['callable'])
        queue = job_msg[0]
        headers = job_msg[1].split(',')
        interval_secs = job_msg[2]
        args = job['args']
        kwargs = job['kwargs']
        class_args = job['class_args']
        class_kwargs = job['class_kwargs']
        unschedule(func, interval_secs=interval_secs, args=args, kwargs=kwargs,
                   class_args=class_args, class_kwargs=class_kwargs,
                   headers=headers, queue=queue)

        print "Removing job ID: {} MSG: {}".format(old_hash, job_msg)
        r.delete(old_hash)
        r.lrem('interval_jobs', 0, old_hash)

        if ':' not in job['path'] and job['callable'] == 'deprovision_inactive_users':  #noqa
            # This is an old broken job
            continue

        jobs.append(job_msg)

    new_interval_job_list = r.lrange('interval_jobs', 0, -1)

    noop = raw_input('# {} jobs in list. Restart the scheduler and press '
                     'Enter to continue '.format(len(new_interval_job_list)))

    dupe_check = []
    for job_msg in jobs:
        job = json.loads(job_msg[3])[1]
        # [u'default', u'guarantee', u'14400', u'["run", {"args": [], "class_args": [7932], "callable": "sync_all_users_and_groups", "kwargs": {}, "path": "integrations.plugins.o365.base:O365", "class_kwargs": {}}]']
        _path = ''
        _class_args = ''
        _callable_args = ''
        if ':' in job['path']:
            if job['class_args']:
                _class_args = ','.join(str(i) for i in job['class_args'])
            if job['class_kwargs']:
                _class_args += ','.join('{}={}'.format(k, v) for k, v in job['class_kwargs'].items())

            _path = '{}({})'.format(job['path'], _class_args)
        else:
            if job['callable'] == 'deprovision_inactive_users':
                # This is an old broken job
                continue
            _path = job['path']

        _path = '{}.{}'.format(_path, job['callable'])

        if job['args']:
            _callable_args = ','.join(str(i) for i in job['args'])
        if job['kwargs']:
            _callable_args = ','.join('{}={}'.format(k, v) for k, v in job['kwargs'].items())

        _path = '{}({})'.format(_path, _callable_args)

        if _path in dupe_check:
            print "! Skipping duplicate job: {}".format(_path)
            continue

        print "# Rescheduling {}".format(_path)
        dupe_check.append(_path)

        func = '.'.join(job['path'], job['callable'])
        queue = job_msg[0]
        headers = job_msg[1].split(',')
        interval_secs = job_msg[2]
        args = job['args']
        kwargs = job['kwargs']
        class_args = job['class_args']
        class_kwargs = job['class_kwargs']
        schedule(func, interval_secs=interval_secs, args=args, kwargs=kwargs,
                 class_args=class_args, class_kwargs=class_kwargs,
                 headers=headers, queue=queue)

print ''
print '# Skipped {} duplicates'.format(len(jobs) - len(dupe_check))
```

com4 commented:
Here is the output from production when I was writing it. 
https://gist.github.com/com4/914c6c5715a0bf0675b5223e8813f83b
